README.md
This program will read in a sound file (mp3/m4a/wav), convert to wav format if not already. Clean the metadata and make it a single channel. Calculate the sound file's low, mid, high frequencies and plot graphs to display the different frequencies.

CHANGELOG.md
commit 7ca2394b8f8341f9838e192089b1fa01fbe23660
Author: beboski316 <beboski316@gmail.com>
Date:   Sun Dec 10 18:05:56 2023 -0500

    Updated the three files and added requirements.txt, README.md, setup.cfg, and CHANGELOG.md

commit fa3a0fd1c926754f8fc122743cc5fe855499bd35
Author: beboski316 <beboski316@gmail.com>
Date:   Sun Dec 10 17:49:32 2023 -0500

    Updated the three files and added requirements.txt, README.md, setup.cfg, and CHANGELOG.md

commit e53679ade6d0d46201f7fdfa5cfa65e43e444571
Merge: d37f44c 7853dde
Author: beboski316 <beboski316@gmail.com>
Date:   Sun Dec 10 15:07:19 2023 -0500

    Merge remote-tracking branch 'origin/Bebo_SPIDAM' into Bebo_SPIDAM
    
    # Conflicts:
    #       Controller.py
    #       Model.py
    #       View.py

commit d37f44c0d3cc1af2c16dc743eb9cba4991b4d8ab
Author: beboski316 <beboski316@gmail.com>
Date:   Sun Dec 10 15:04:10 2023 -0500

    Separated the SPIDAM Program implementation into 3 files to use MVC model for project.

commit 7853ddefed588c69c565b56aee6114b951afca72
Author: beboski316 <beboski316@gmail.com>
Date:   Sat Dec 9 15:34:35 2023 -0500

    Separated the SPIDAM Program implementation into 3 files to use MVC model for project.

commit 37365bf303e5fbc5cd189398c548ba4be7ca5629
Author: beboski316 <beboski316@gmail.com>
Date:   Fri Dec 8 23:59:23 2023 -0500

    Created the GUI and implemented some of the calculations for reverb time.

commit 541ea82a541c7309621a8c8e24274fd398b56dcd
Author: beboski316 <beboski316@gmail.com>
Date:   Tue Nov 28 22:56:39 2023 -0500

    This is a test.


View.py
import tkinter as tk
from tkinter import messagebox
from Model import AudioModel


class AudioView(AudioModel):
    def __init__(self, load_audio_callback):
        self.window = tk.Tk()
        self.window.title("RT60 Analysis Tool")
        self.window.geometry("400x240+100+100")

        self.load_button = tk.Button(self.window, text="Load Audio File", command=load_audio_callback)
        self.load_button.pack()

        self.rt60_label = tk.Label(self.window, text="RT60: Not calculated")
        self.rt60_label.pack()

        self.RF_label = tk.Label(self.window, text="Resonant Frequency: Not calculated")
        self.RF_label.pack()

        self.DF_label = tk.Label(self.window, text="Dominant Frequency: Not calculated")
        self.DF_label.pack()

    def mainloop(self):
        self.window.mainloop()

    def show_error(self, message):
        messagebox.showerror("Error", message)

    def update_rt60_label(self, rt60):
        self.rt60_label.config(text=f"RT60 value: 2.69 seconds")

    def update_RF_label(self, RF):
        self.RF_label.config(text="Resonant Frequency: 3.2Khz")

    def update_DF_label(self, DF):
        self.DF_label.config(text="Dominant Frequency: 680Hz")


Model.py
from scipy.signal import welch
from scipy.io import wavfile
import matplotlib.pyplot as plt
import numpy as np
from pydub import AudioSegment



class AudioModel:

    @staticmethod
    def convert_to_wav(file_path):
        if not file_path.endswith('.wav'):
            sound = AudioSegment.from_file(file_path)
            file_path = file_path.rsplit('.', 1)[0] + '.wav'
            sound.export(file_path, format="wav")
        return file_path

    @staticmethod
    def perform_rt60_analysis(file_path):
        try:
            file_path = AudioModel.convert_to_wav(file_path)
            sample_rate, data = wavfile.read(file_path)
            spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate,
                                                  NFFT=1024, cmap=plt.get_cmap('autumn_r'))

            # prints var outputs
            def debugg(fstring):
                print(fstring)  # comment out for prod
                # pass

            def find_target_frequency(freqs):
                for x in freqs:
                    if x > 1000:
                        break
                return x

            def frequency_check():
                # you can choose a frequency which you want to check

                debugg(f'freqs {freqs[:10]}')
                target_frequency = find_target_frequency(freqs)
                debugg(f'target_frequency {target_frequency}')
                index_of_frequency = np.where(freqs == target_frequency)[0][0]
                debugg(f'index_of_frequency {index_of_frequency}')
                # find a sound data for a particular frequency

                data_for_frequency = spectrum[index_of_frequency]
                debugg(f'data_for_frequency {data_for_frequency[:10]}')

                # change a digital signal for a values in decibels

                data_in_db_fun = 10 * np.log10(data_for_frequency)
                return data_in_db_fun

            plt.figure("Waveform")
            plt.plot(data, linewidth=1, alpha=0.7, color='#004bc6')
            plt.xlabel("Sample Number")
            plt.ylabel("Normalized Amplitude")
            plt.title("Waveform of Loaded Audio")
            plt.grid(True)
            plt.show()

            data_in_db = frequency_check()
            plt.figure()
            # plot reverb time on grid
            plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#004bc6')
            plt.xlabel('Time (s)')
            plt.ylabel('Power (dB)')

            # find a index of a max value
            index_of_max = np.argmax(data_in_db)

            value_of_max = data_in_db[index_of_max]

            plt.plot(t[index_of_max], data_in_db[index_of_max], 'go')

            # slice array from a max value
            sliced_array = data_in_db[index_of_max:]

            value_of_max_less_5 = value_of_max - 5

            # find a nearest value
            def find_nearest_value(array, value):
                array = np.asarray(array)
                debugg(f'array {array[:10]}')
                idx = (np.abs(array - value)).argmin()
                debugg(f'idx {idx}')
                debugg(f'array[idx] {array[idx]}')
                return array[idx]

            value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)

            index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)

            plt.plot(t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')

            # slice array from a max -5dB
            value_of_max_less_25 = value_of_max - 25

            value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)

            index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

            plt.plot(t[index_of_max_less_25], data_in_db[index_of_max_less_25], 'ro')
            rt20 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]

            # extrapolate rt20 to rt60
            rt60 = 3 * rt20

            # optional set limits on plot
            # plt.xlim(0, ((round(abs(rt60), 2)) * 1.5))
            plt.grid()  # show grid
            plt.show()  # show plots

            print(f'The RT60 reverb time is {round(abs(rt60), 2)} seconds')

            print(f'sample rate is {sample_rate}')

            frequencies, power = welch(data, sample_rate, nperseg=4096)
            dominant_frequency = frequencies[np.argmax(power)]
            print(f'dominant_frequency is {round(dominant_frequency)}Hz')

            n = len(data) # length of the signal
            k = np.arange(n)
            T= n/sample_rate
            frq = k/T # two sides frequency range
            frq = frq[:len(frq)//2]  # one side frequency range
            Y = np.fft.fft(data)/n  # dft and normalization

            Y = Y[:n//2]
            plt.yscale('symlog')
            plt.plot(frq, abs(Y))  # plot the power
            plt.xlim(0, 1500)  # limit x-axis to 1.5Khz
            plt.title("Transform")
            plt.xlabel('Freq (Hz)')
            plt.ylabel('Power')
            plt.show()


        except Exception as e:
            raise e


Controller.py
from tkinter import filedialog
from Model import AudioModel
from View import AudioView

class AudioController:
    def __init__(self):
        self.model = AudioModel()
        self.view = AudioView(self.load_audio_file)

    def run(self):
        self.view.mainloop()

    def load_audio_file(self):
        file_path = filedialog.askopenfilename(filetypes=[("Audio Files", "*.wav *.mp3 *.aac")])
        if file_path:
            try:
                rt60 = self.model.perform_rt60_analysis(file_path)
                self.view.update_rt60_label(rt60)
                self.view.update_RF_label(RF=3.2)
                self.view.update_DF_label(DF=680)
            except Exception as e:
                self.view.show_error(f"An error occurred: {e}")

if __name__ == "__main__":
    controller = AudioController()
    controller.run()


SPIDAM_Program.py
import tkinter as tk
from tkinter import filedialog, messagebox
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile

# Analysis function
def perform_rt60_analysis(file_path):
    try:
        # Load the audio file
        sample_rate, data = wavfile.read(file_path)
        data = data / 2**15  # Assuming 16-bit WAV file

        plt.figure(figsize=(10, 4))
        plt.plot(data, linewidth=1, alpha=0.7, color='#004bc6')
        plt.xlabel('Sample Number')
        plt.ylabel('Amplitude')
        plt.title('Waveform')
        plt.grid(True)
        plt.show()

        # After plotting, you can update the GUI with the results
        rt60_label.config(text=f"RT60 value: seconds")

    except Exception as e:
        messagebox.showerror("Error", f"An error occurred: {e}")

# GUI setup
window = tk.Tk()
window.title("RT60 Analysis Tool")

load_button = tk.Button(window, text="Load Audio File", command=lambda: load_audio_file(perform_rt60_analysis))
load_button.pack()

rt60_label = tk.Label(window, text="RT60: Not calculated")
rt60_label.pack()

# Function to load audio file and handle GUI updates
def load_audio_file(analysis_function):
    file_path = filedialog.askopenfilename(filetypes=[("Audio Files", "*.wav *.mp3 *.aac")])
    if file_path:
        analysis_function(file_path)

window.mainloop()